lib/charms/alertmanager_k8s/v0/alertmanager_remote_configuration.py

```diff
@@ -0,0 +1,454 @@
+        self._charm = charm
+        self.alertmanager_config = alertmanager_config
+        self._relation_name = relation_name
+
+        on_relation = self._charm.on[self._relation_name]
+
+        self.framework.observe(on_relation.relation_joined, self._on_relation_joined)
+
+    @classmethod
+    def with_config_file(
+        cls,
+        charm: CharmBase,
+        config_file: Path,
+        relation_name: str = DEFAULT_RELATION_NAME,
+    ):
+        """The RemoteConfigurationProvider object factory.
+
+        This factory provides an alternative way of instantiating the RemoteConfigurationProvider.
+        While the default constructor requires passing a config dict, the factory allows using
+        a configuration file path.
+
+        Args:
+            charm: The charm object that instantiated this class.
+            config_file: Path to the Alertmanager configuration file.
+            relation_name: Name of the relation with the `alertmanager_remote_configuration`
+                interface as defined in metadata.yaml. Defaults to `remote-configuration`.
+
+        Returns:
+            RemoteConfigurationProvider object
+        """
+        return cls(charm, cls.load_config_file(config_file), relation_name)
+
+    def _on_relation_joined(self, _) -> None:
+        """Event handler for RelationJoinedEvent.
+
+        Takes care of pushing Alertmanager configuration to the relation data bag.
+        """
+        if not self._charm.unit.is_leader():
+            return
+        if self.alertmanager_config:
+            self.update_relation_data_bag(self.alertmanager_config)
+        else:
+            logger.warning("Alertmanager configuration not available. Ignoring...")
+
+    @staticmethod
+    def load_config_file(path: Path) -> dict:
+        """Reads given Alertmanager configuration file and turns it into a dictionary.
+
+        Args:
+            path: Path to the Alertmanager configuration file
+
+        Returns:
+            dict: Alertmanager configuration file in a form of a dictionary
+
+        Raises:
+            ConfigReadError: if a problem with reading given config file happens
+        """
+        try:
+            with open(path, "r") as config_yaml:
+                config = yaml.safe_load(config_yaml)
+            return config
+        except (FileNotFoundError, OSError, yaml.YAMLError) as e:
+            raise ConfigReadError(path) from e
+
+    def update_relation_data_bag(self, alertmanager_config: dict) -> None:
+        """Updates relation data bag with Alertmanager config and templates.
+
+        Before updating relation data bag, basic sanity check of given configuration is done.
+
+        Args:
+            alertmanager_config: Alertmanager configuration dictionary.
+        """
+        config = alertmanager_config
+        templates = self._get_templates(config)
+        for relation in self._charm.model.relations[self._relation_name]:
+            if config_main_keys_are_valid(config):
```

Comment:
We can probably swap the two around?
```suggestion
        if config_main_keys_are_valid(config):
            for relation in self._charm.model.relations[self._relation_name]:
```
- What do we do if keys fails validation?
- Do we need a leader guard?

Reply:
Indeed, why should we bother if config is invalid. 
I've added a logger (actually moved it from the check you mentioned below) to inform about attempt of pushing invalid config.
Good catch regarding the leader. Added as well. Thanks!